jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java

```diff
@@ -437,6 +437,36 @@ public CaseDefinition getCase(String deploymentId, String caseDefinitionId) {
         return processInstances;
     }
 
+    @Override
+    public Collection<CaseInstance> getSubCaseInstancesByParentCaseId(String parentCaseId, List<CaseStatus> statuses, QueryContext queryContext) {
+        Map<String, Object> params = new HashMap<String, Object>();
+        params.put("parentCaseId", Arrays.asList(parentCaseId));
+        params.put("statuses", resolveCaseStatuses(statuses));
+        applyQueryContext(params, queryContext);
+        applyDeploymentFilter(params);
+        List<CaseInstance> processInstances = commandService.execute(new QueryNameCommand<List<CaseInstance>>("getSubCaseInstancesByParentCaseIds", params));
+
+        return processInstances;
+    }
+
+    @Override
+    public Collection<CaseInstance> getAllDescendantSubCaseInstancesByParentCaseId(String parentCaseId, List<CaseStatus> statuses, QueryContext queryContext) {
```

Comment:
Here the QueryContext is applied per level:
Suppose limit is X, then if each level returns X-1 results, the API will return N levels * (X-1) results, exceeding the global limit (X). 
Same issue happens with the offset: if first level subcases number is lower than the offset, then it won't iterate over the next level (though there're remaining results).
Perhaps for this API, it's better an implementation without queryContext, querying as many times as needed per level and consolidate them all to be returned in a single call.
What do you think?

Reply:
done